modules/currency-l0/src/main/scala/org/tessellation/currency/l0/snapshot/services/StateChannelBinarySender.scala

```diff
@@ -0,0 +1,238 @@
+  binary: Hashed[StateChannelSnapshotBinary],
+  enqueuedAtOrdinal: SnapshotOrdinal,
+  sendsSoFar: NonNegLong
+)
+
+case class State(
+  pending: Queue[TrackedBinary],
+  cap: NonNegLong,
+  retryMode: Boolean,
+  noConfirmationsSinceRetryCount: NonNegLong,
+  backoffExponent: NonNegLong
+)
+
+object State {
+  def empty: State = State(
+    pending = Queue.empty[TrackedBinary],
+    cap = 4L,
+    retryMode = false,
+    noConfirmationsSinceRetryCount = 0L,
+    backoffExponent = 0L
+  )
+}
+
+trait StateChannelBinarySender[F[_]] {
+  def processPending: F[Unit]
+  def clearPending: F[Unit]
+  def confirm(globalSnapshot: GlobalIncrementalSnapshot): F[Unit]
+  def process(binaryHashed: Hashed[StateChannelSnapshotBinary]): F[Unit]
+}
+
+object StateChannelBinarySender {
+  def make[F[_]: Async: Hasher](
+    identifierStorage: IdentifierStorage[F],
+    globalL0ClusterStorage: L0ClusterStorage[F],
+    lastGlobalSnapshotStorage: LastSnapshotStorage[F, GlobalIncrementalSnapshot, GlobalSnapshotInfo],
+    stateChannelSnapshotClient: StateChannelSnapshotClient[F]
+  ): F[StateChannelBinarySender[F]] =
+    Ref
+      .of[F, State](State.empty)
+      .map(make[F](_, identifierStorage, globalL0ClusterStorage, lastGlobalSnapshotStorage, stateChannelSnapshotClient))
+
+  def make[F[_]: Async: Hasher](
+    stateR: Ref[F, State],
+    identifierStorage: IdentifierStorage[F],
+    globalL0ClusterStorage: L0ClusterStorage[F],
+    lastGlobalSnapshotStorage: LastSnapshotStorage[F, GlobalIncrementalSnapshot, GlobalSnapshotInfo],
+    stateChannelSnapshotClient: StateChannelSnapshotClient[F]
+  ): StateChannelBinarySender[F] =
+    new StateChannelBinarySender[F] {
+
+      private val noConfirmationsToTriggerRetryMode: PosLong = 5L
+      private val confirmedCountMultiplier: PosLong = 4L
+
+      private val logger = Slf4jLogger.getLogger
+
+      def process(binary: Hashed[StateChannelSnapshotBinary]): F[Unit] =
+        lastGlobalSnapshotStorage.getOrdinal.map(_.getOrElse(SnapshotOrdinal.MinValue)).flatMap { currentOrdinal =>
+          stateR.modify {
+            case state @ State(pending, _, retryMode, _, _) =>
+              val updatedPending = pending :+ TrackedBinary(binary, enqueuedAtOrdinal = currentOrdinal, 0L)
+              val action =
+                if (retryMode) logger.warn(s"[RetryMode] Snapshot binary of hash ${binary.hash} enqueued.")
+                else
+                  post(binary).flatTap(_ => logger.info(s"Snapshot binary of hash ${binary.hash} enqueued and sent to GL0"))
+              val newState = state.copy(pending = updatedPending)
+              (newState, action)
+          }.flatMap(identity)
+        }
+
+      def confirm(globalSnapshot: GlobalIncrementalSnapshot): F[Unit] = for {
+        identifier <- identifierStorage.get
+        binaries = globalSnapshot.stateChannelSnapshots.get(identifier).toList.flatMap(_.toList)
+        confirmedHashesInGlobalSnapshot <- binaries.traverse(_.toHashed).map(_.map(_.hash))
```

Comment:
Since this is searched inside a loop over the pending queue it should be Set, unless the queue is always going to be very small

Reply:
The `stateChannelSnapshots` for `identifier` are `NonEmptyList` but all these snapshots are unique for sure (I guess we didn't use `NonEmptySet` because it is `SortedSet` under the hood so it reorders binaries but we need to keep the order). I can make it a `Set` just to be sure here though.

Reply:
And of course `contains` on `Set` is much faster if this is what you mean.